cmd/variables.go

```diff
@@ -0,0 +1,288 @@
+		http.MethodGet,
+		"/ui/variables",
+		nil,
+		"application/json",
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var escs []EnvSpecificConfig
+	if err := json.NewDecoder(*result).Decode(&escs); err != nil { return nil, err }
+
+	if mapDsName {
+		dsmap, err := getDataServices(api)
+		if err != nil { return nil, err }
+		for i, esc := range escs {
+			if esc.DataServiceID != "" {
+				if esc.DataServiceID == fakeDefaultDataServiceId {
+					esc.DataServiceName = "default"
+					escs[i] = esc
+				} else if ds, ok := dsmap[esc.DataServiceID]; ok {
+						esc.DataServiceName = ds.Name
+						escs[i] = esc
+				}
+			}
+		}
+	}
+
+	if verbose {
+		fmt.Println(text.SuccessWithTime("Success getting variable values", escStart))
+	}
+	return escs, nil
+}
+
+// setvarCmd represents the setvariable command
+var setvarCmd = &cobra.Command{
+	Use:   "setvariable",
+	Short: "Set a Skuid environment variable.",
+	Long:  "Set a Skuid envorinment variable. These can be used as credentials for a data source without revealing them.",
+	Run: func(cmd *cobra.Command, args []string) {
+
+		fmt.Println(text.RunCommand("Set Variable"))
+
+		api, err := platform.Login(
+			host,
+			username,
+			password,
+			apiVersion,
+			metadataServiceProxy,
+			dataServiceProxy,
+			verbose,
+		)
+
+		if err != nil {
+			fmt.Println(text.PrettyError("Error logging in to Skuid site", err))
+			os.Exit(1)
+		}
+
+		variableStart := time.Now()
+		err = setEsc(api)
+		if err != nil {
+			fmt.Println(text.PrettyError("Error setting variable in Skuid site", err))
+			os.Exit(1)
+		}
+		successMessage := "Successfully set variable in Skuid Site"
+		if verbose {
+			fmt.Println(text.SuccessWithTime(successMessage, variableStart))
+		} else {
+			fmt.Println(successMessage + ".")
+		}
+	},
+}
+
+func setEsc(api *platform.RestApi) error {
+	if verbose {
+		fmt.Println(text.VerboseSection("Setting Variable"))
+	}
+	if variablename == "" {
+		return errors.New("Variable name is required for this command.")
+	}
+	if variablevalue == "" {
+		fmt.Println("Enter value:")
+		valbytes, err := terminal.ReadPassword(0)
+		if err != nil { return errors.New("Error reading value from prompt.") }
+		variablevalue = string(valbytes)
+	}
+	api.Connection.APIVersion = "1"
+
+	dataServiceId := ""
+	if variabledataservice != "" {
+		if _, err := uuid.FromString(variabledataservice); err == nil {
+			dataServiceId = variabledataservice
+		} else if variabledataservice == "default" {
+			dataServiceId = fakeDefaultDataServiceId
+		} else {
+			dataservices, err := getDataServices(api)
+			if err != nil {
+				return err
+			}
+			found := false
+			for _, ds := range dataservices {
+				if ds.Name == variabledataservice {
+					found = true
+					dataServiceId = ds.ID
+				}
+			}
+			if !found {
+				return errors.New("Could not find specified Data Service by name.")
+			}
+		}
+	}
+	body := map[string]interface{}{}
+	body["name"] = variablename
+	body["value"] = variablevalue
+	if dataServiceId != "" {
```

Comment:
I think we should always populate `dataServiceId` , but as above if it is "" we should default it to the default data service id, otherwise we'll have variables not associated with any data service. i tried this and the result was wonky.